src/client/common/application/customEditorService.ts

```diff
@@ -48,35 +49,54 @@ export class CustomEditorService implements ICustomEditorService, IExtensionSing
         let updateType: 'added' | 'removed' | undefined;
         try {
             const settings = this.workspace.getConfiguration('workbench', undefined);
-            const editorAssociations = settings.get('editorAssociations') as {
-                viewType: string;
-                filenamePattern: string;
-            }[];
+            const editorAssociations = settings.get('editorAssociations') ?? {}; // At this point we don't know if this is the old or new format
 
-            // Update the settings.
-            if (
-                this.useCustomEditorApi &&
-                (editorAssociations.length === 0 || !editorAssociations.find((item) => item.viewType === ViewType))
-            ) {
-                editorAssociations.push({
-                    viewType: ViewType,
-                    filenamePattern: '*.ipynb'
-                });
-                updateType = 'added';
-                await settings.update('editorAssociations', editorAssociations, vscode.ConfigurationTarget.Global);
-                sendTelemetryEvent(Telemetry.UpdateCustomEditorAssociation, undefined, { type: 'added' });
-            }
+            if (Array.isArray(editorAssociations)) {
+                // Old format for editorAssociations setting
+                const hasCustomEditorAssociation =
+                    editorAssociations.find((item) => item.viewType === ViewType) !== undefined;
+
+                // Update the settings.
+                if (this.useCustomEditorApi && (editorAssociations.length === 0 || !hasCustomEditorAssociation)) {
+                    const updatedSettings = [
+                        ...editorAssociations,
+                        {
+                            viewType: ViewType,
+                            filenamePattern: '*.ipynb'
+                        }
+                    ];
+                    updateType = 'added';
+                    await settings.update('editorAssociations', updatedSettings, vscode.ConfigurationTarget.Global);
+                    sendTelemetryEvent(Telemetry.UpdateCustomEditorAssociation, undefined, { type: 'added' });
+                }
+
+                // Revert the settings.
+                if (!this.useCustomEditorApi && hasCustomEditorAssociation) {
+                    const updatedSettings = editorAssociations.filter((item) => item.viewType !== ViewType);
+                    updateType = 'removed';
+                    await settings.update('editorAssociations', updatedSettings, vscode.ConfigurationTarget.Global);
```

Comment:
Seems like these two lines "await settings.update(...)" and "sendTelemetryEvent(Telemetry...)" could be pulled out and called just once throughout this whole function. Not a huge deal though.

Reply:
I don't believe we want to unconditionally send the telemetry event or update the settings object outside of these four cases. For example, if we should be using custom editor and the association is already present, the settings object hasn't changed so shouldn't be updated, and the telemetry event represents a settings update and no update has taken place so shouldn't be sent.